### ISSUE TITLE
[LUA] enhancements

### ISSUE THREAD

**Issue by Spikeone** [Return-To-The-Roots/s25client#36]
Save Lua State - Save the Lua state to savegames so you can load and continue
Increase Lua filesize - currently 8192b which isn't enough for making enhanced scripts

Take ware - you should be able to wares of a player
Add ware to building / Take ware from building - adds/takes wares from a specific building (includes non store houses)
Set addons - control addon settings
Set options - control options
Post message with picture (also picture and location) - give an id for an image displayed
Mission statement with picture - same as post message with picture but for mission statement
Add Ship - adds a ship at given position for given player
Set ship dest - sets a ships destination (by ship and player id)
Set ship wares - sets wares of a ship by given id
Destroy ship - destroy a ship by given player and ship Id, add fire at position
Add building at - adds a building at a given position for given player ( if military building also set soldiers and rank)
Destroy building at - destroys a building at a position
Building finished - called when a building was finished, gives position, player id, type
Animate bob - takes any moved bob ingame, set way by coordinates, player by id, movement types: circle (if last waypoint, start move to first and restart(not teleport)), patrol (if last, restart in reversed order(last =first, first =last), strop and inform (informs the script that an animated bob reached its destination)
Show information - adds up to n info texts in the upper right corner (eg for counters, short quest text)
Set relation - set relation between players/ai by ID and type
Get military strength - returns military strength of player by id

-----
I'll continue this list when I'm home, mobile phones are not made for such tasks ;)

**Comment by Flamefire** [comment]
Some of those are quite useful. E.g. save/load and increase lua size (currently there is probably even a bug, when the lua size exceeds tcp packet size...)
However note that making those functions adds quite some complexity and using it in multiplayer games requires careful design by the script creator or there will be asyncs. I think, if the script purely depends on events and GF numbers, then even adding wares would work without asyncs, but I'm not sure.

I'd make save/load, lua size and option settings (player faces, addon settings etc) top priority as this is required for good missions. Maybe easiest: Ask the player, if he wants to start map as mission, and if he does, make most things readonly and call the lua script.

**Comment by jhkl** [comment]
> However note that making those functions adds quite some complexity and using it in multiplayer games requires careful design by the script creator or there will be asyncs.

All gameplay related functions should be realised as GameCommand or do I miss something here? Then there should be no (additional) danger for asyncs, right?

**Comment by Spikeone** [comment]
Adding wares did not create any async (maqs said something like "when every player has the same script, there can't occur an async). Idont get how this is mapcreator related, if a function is not multiplayer compatible (and will never be) the game shouldn't start at all.

**Comment by Flamefire** [comment]
Maqs is right. If the same script runs on every client, it is fine as long as nothing player-dependant is done. The script is synchronized as long as only events based on the game itself are used. This includes GameFrames, discovered territory, actually placed buildings or orders (opposite to the intermediate showing of a buildings placement or e.g. change coins visual indicator) but *excludes* player dialogues, pc specific stuff (local time, files,...) or direct influence of player actions (clicking somewhere,...)

So I would not add any such function to lua or if they are added, prohibit them in multiplayer matches. This is possible with the C++ wrapper by simply not exporting those functions in multiplayer mode. However this would result in errors during the game if a map maker did use them...

**Comment by Spikeone** [comment]
Register player clicks and add some basic input should also work in multiplayer.

Something like vbs should be enough. For example one could add a bridge builder where. You could purchase a bridge. If any player clicks on that building a window should open open up asking the player if he wants to buy a bridge, if 1 is returned an other window opens with a simple input asking for x,y, length as example. As the result is for every player the same this shouldn't async. How this could work - I don't know. But I still think that this would be great (eg for shops or trade posts).

Maybe one could add objects so every player bows how that object behaves?

**Comment by Flamefire** [comment]
Nope. This would not be the same for all players. Player clicks are player-only, clicks are not sent to other players so no player will know, if he bought the bridge. This would need to be sent via a GameCommand. But then we would need 2 functions for each action: One that sends a GC and one that does not. Also adding GCs is a bit work...

I also think, if you want to add shops etc: Use the WC3 editor and import some textures ;) This is SPARTA! (aehm, settlers 2 :D)

**Comment by Spikeone** [comment]
Point is interaction. Without player interaction it would be quite sad. I still think that it is a good thing we should add if we could.

If it's hard to add... Mhh.
